core/modules/replica/GetReplicasQservMgtRequest.cc

```diff
@@ -0,0 +1,179 @@
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/GetReplicasQservMgtRequest.h"
+
+// System headers
+#include <future>
+#include <set>
+#include <stdexcept>
+
+// Third party headers
+#include "XrdSsi/XrdSsiProvider.hh"
+#include "XrdSsi/XrdSsiService.hh"
+
+// Qserv headers
+#include "global/ResourceUnit.h"
+#include "lsst/log/Log.h"
+#include "replica/Configuration.h"
+#include "replica/ServiceProvider.h"
+
+// This macro to appear witin each block which requires thread safety
+#define LOCK_GUARD std::lock_guard<std::mutex> lock(_mtx)
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.GetReplicasQservMgtRequest");
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+GetReplicasQservMgtRequest::pointer GetReplicasQservMgtRequest::create(
+                                        ServiceProvider::pointer const& serviceProvider,
+                                        boost::asio::io_service& io_service,
+                                        std::string const& worker,
+                                        std::string const& databaseFamily,
+                                        bool inUseOnly,
+                                        GetReplicasQservMgtRequest::callback_type onFinish) {
+    return GetReplicasQservMgtRequest::pointer(
+        new GetReplicasQservMgtRequest(serviceProvider,
+                                       io_service,
+                                       worker,
+                                       databaseFamily,
+                                       inUseOnly,
+                                       onFinish));
+ }
+
+GetReplicasQservMgtRequest::GetReplicasQservMgtRequest(
+                                ServiceProvider::pointer const& serviceProvider,
+                                boost::asio::io_service& io_service,
+                                std::string const& worker,
+                                std::string const& databaseFamily,
+                                bool inUseOnly,
+                                GetReplicasQservMgtRequest::callback_type onFinish)
+    :   QservMgtRequest(serviceProvider,
+                        io_service,
+                        "QSERV:GET_REPLICAS",
+                        worker),
+        _databaseFamily(databaseFamily),
+        _inUseOnly(inUseOnly),
+        _onFinish(onFinish),
+        _qservRequest(nullptr) {
+
+    if (not _serviceProvider->config()->isKnownDatabaseFamily(_databaseFamily)) {
+        throw std::logic_error(
+                "GetReplicasQservMgtRequest: unknown database family: " +
+                _databaseFamily);
+    }
+}
+
+QservReplicaCollection const& GetReplicasQservMgtRequest::replicas() const {
+    if (not ((_state == State::FINISHED) and (_extendedState == ExtendedState::SUCCESS))) {
```

Comment:
_state and _extendedState should have some sort of protection, probably using the same mutex so their state doesn't get skewed.